tests/test_directives.py

```diff
@@ -0,0 +1,180 @@
+
+@pytest.fixture()
+def available_directives_list():
+    return [NP, NRANKS, NGPU, OMP_NUM_THREADS,
+            EXECUTABLE, WALLTIME, MEMORY,
+            PROCESS_FRACTION]
+
+
+class TestItems:
+    """
+    Tests for _DirectivesItem class
+    """
+    def test_default(self):
+        assert NP.default == 1
+        assert NGPU.default == 0
+        assert NRANKS.default == 0
+        assert OMP_NUM_THREADS.default == 0
+        assert EXECUTABLE.default == sys.executable
+        assert WALLTIME.default == 12.0
+        assert MEMORY.default == 4
+        assert PROCESS_FRACTION.default == 1.
+
+    def test_validation(self, available_directives_list):
+        for directive in available_directives_list:
+            directive.validation(directive.default)
+
+    # def test_validation_none(self):
+    #     available_directives_list =
+    #     for directive in available_directives_list:
+    #         directive.validation(directive.default)
+
+    def test_validation_manual(self):
+        directive_copy = deepcopy(NP)
+
+        def div_by_ten(v):
+            if v % 10 == 0:
+                return True
+            return False
+
+        directive_copy.validation = div_by_ten
+        assert directive_copy.validation(10)
+        assert not directive_copy.validation(5)
+        with pytest.raises(TypeError):
+            directive_copy.validation('4')
+
+    def test_serial(self):
+        assert NP.serial(4, 2) == 4
+        assert NRANKS.serial(4, 2) == 4
+        assert NGPU.serial(4, 2) == 4
+        assert OMP_NUM_THREADS(4, 2) == 4
+        assert EXECUTABLE.serial('Path1', 'Path2') == 'Path1'
+        assert WALLTIME.serial(4, 2) == 6
+        assert MEMORY.serial(4, 2) == 4
+        assert PROCESS_FRACTION(4.0, 2.0) == 4.0
+
+    def test_parallel(self):
+        assert NP.parallel(4, 2) == 6
+        assert NRANKS.parallel(4, 2) == 6
+        assert NGPU.parallel(4, 2) == 6
+        assert OMP_NUM_THREADS(4, 2) == 6
+        assert EXECUTABLE.parallel('Path1', 'Path2') == 'Path1'
+        assert WALLTIME.parallel(4, 2) == 4
+        assert MEMORY.parallel(4, 2) == 6
+        assert PROCESS_FRACTION.parallel(4.0, 2.0) == 4.0
+
+    def test_parallel_manual(self):
+        directive_copy = deepcopy(NP)
+        directive_copy.parallel = lambda x, y: x*y
+        parallel = directive_copy.parallel(3, 2)
+        assert parallel == 6
+        directive_copy.validation(parallel)
+        with pytest.raises(TypeError):
+            directive_copy.parallel('4', '2')
+
+    def test_finalize(self):
+        dict_directives = {'nranks': NRANKS.default, 'omp_num_threads': OMP_NUM_THREADS.default}
+        assert NP.finalize(2, dict_directives) == 2
+        dict_directives['nranks'] = 2
+        dict_directives['omp_num_threads'] = 4
+        assert NP.finalize(2, dict_directives) == 8
+        dict_directives['nranks'] = NRANKS
+        assert NP.finalize(2, dict_directives) == 2
+
+    def test_manual_item(self, setUp):
```

Comment:
This test a bunch of the functionality I wanted; however, it test them in one test. It would be nicer to have individual tests for the individual functionality. Also, `PRODUCT` does not need to be capitalized.

Reply:
Noted